script/release

```diff
@@ -1,43 +1,165 @@
-
-echo "####################################################"
-echo "Now, open a PR with this branch and merge it to main"
-echo "Then, run bundle exec rake release on main to release the gem to RubyGems"
-echo "Finally, create a GitHub release https://github.com/primer/view_components/releases/new with the changes from CHANGELOG"
-echo "####################################################"
+fetch() {
+  git fetch --all
+}
+
+branch_name() {
+  git symbolic-ref --short HEAD
+}
+
+local_history_is_clean() {
+  history=$(git rev-list --count --right-only @{u}...HEAD)
+  [ "$history" == "0" ]
+}
+
+remote_history_is_clean() {
+  history=$(git rev-list --count --left-only @{u}...HEAD)
+  [ "$history" == "0" ]
+}
+
+tag_exists_on_remote() {
+  git rev-parse --quiet --verify refs/tags/v$1.$2.$3 > /dev/null
+}
+
+working_tree_is_clean() {
+  status=$(git status --porcelain)
+  [ "$status" == "" ]
+}
+
+create_release_branch() {
+  git checkout -B release
+}
+
+update_readme() {
+  sed -i '' "/## main/ {a\
+  \\
+  \\
+  ## $1.$2.$3
+  }" CHANGELOG.md
+}
+
+update_ruby_version() {
+  # Update version file
+  sed -E -i '' \
+      -e "s/MAJOR = [0-9]+/MAJOR = $1/g" \
+      -e "s/MINOR = [0-9]+/MINOR = $2/g" \
+      -e "s/PATCH = [0-9]+/PATCH = $3/g" \
+      lib/primer/view_components/version.rb
+}
+
+update_gemfiles() {
+  # Update Gemfile.lock
+  bundle
+  pushd demo
+  bundle
+  popd
+}
+
+
+update_npm() {
+  npm version --no-git-tag-version "$1.$2.$3"
+  yarn
+}
+
+add_changed_files() {
+  git add \
+    CHANGELOG.md \
+    Gemfile.lock \
+    demo/Gemfile.lock \
+    lib/primer/view_components/version.rb \
+    app/assets/javascripts \
+    package.json \
+    yarn.lock
+}
+
+commit() {
+  git commit -m "release $1.$2.$3"
+}
+
+push() {
+  git push origin release
+
+  echo "####################################################"
+  echo "Now, open a PR with this branch and merge it to main"
+  echo "Then, run script/publish on main to release the gem and the NPM package"
+  echo "Finally, create a GitHub release https://github.com/primer/view_components/releases/new with the changes from CHANGELOG"
+  echo "####################################################"
+}
+
+main() {
+  version=$(ruby ./lib/primer/view_components/version.rb)
+  version=(${version//./ })
+  major=${version[0]}
+  minor=${version[1]}
+  patch=${version[2]}
+
+  echo "==================="
+  echo "Prerequisite Checks"
+  echo "==================="
+
+  if ! working_tree_is_clean; then 
+    echo "Error: unclean working tree"
+    exit 1
+  fi 
+
+  if [ "$(branch_name)" !== "main" ]; then
+    echo "Error: can only make a release on the main branch"
+    exit 1
+  fi
+
+  fetch
+  if ! remote_history_is_clean; then 
+    echo "Error: changes exist on origin not pulled into this branch. Please pull"
+    exit 1
+  fi 
+
+  if ! local_history_is_clean; then 
+    echo "Error: changes exist that haven't been pushed to origin. Please pull"
+    exit 1
+  fi 
```

Comment:
So it now checks:

 - You're on `main`
 - You have a clean tree, no erroneous files
 - You have pushed up everything
 - You've pulled down everything
 - You're not going to overwrite an existing tag

I've tried to test these as best I can but I guess we'll try running this for real when merged and see how it goes ;)

Reply:
Having some errors in this script is fine, we can start using it and fix it as it goes